tests/periph_pwm/main.c

```diff
@@ -20,42 +20,135 @@
  * every 1s on every channel.
  *
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Semjon Kerner <semjon.kerner@fu-berlin.de>
  *
  * @}
  */
 
 #include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
 
 #include "xtimer.h"
+#include "shell.h"
 #include "timex.h"
 #include "periph/pwm.h"
 
-#define INTERVAL    (10LU * US_PER_MS) /* 10 ms */
-#define STEP        (10)
+#define OSC_INTERVAL    (10LU * US_PER_MS) /* 10 ms */
+#define OSC_STEP        (10)
+#define OSC_MODE        PWM_LEFT
+#define OSC_FREQU       (1000U)
+#define OSC_STEPS       (1000U)
+#define PWR_SLEEP       (1U)
 
-#define MODE        PWM_LEFT
-#define FREQU       (1000U)
-#define STEPS       (1000U)
+static uint32_t initiated;
 
-
-int main(void)
+static unsigned _get_dev(const char *dev_str)
 {
+    unsigned dev = (unsigned)atoi(dev_str);
+    if (dev >= PWM_NUMOF) {
+        printf("[pwm] Error: device PWM_DEV(%u) is unknown\n", dev);
+        return UINT_MAX;
+    }
+    return dev;
+}
+
+int init(int argc, char** argv) {
+
+    if (argc != 5) {
+        printf("usage: %s <dev> <mode> <frequency> <resolution>\n", argv[0]);
+        printf("\tdev: device by number between 0 and %d\n", PWM_NUMOF - 1);
+        puts("\tmode:\n");
+        puts("\t\t0: left aligned\n");
+        puts("\t\t1: right aligned\n");
+        puts("\t\t2: center aligned\n");
+        puts("\tfrequency: desired frequency in Hz\n");
+        printf("\tresolution: number between 2 and %d\n", UINT16_MAX);
+        return 1;
+    }
+
+    unsigned dev = _get_dev(argv[1]);
+    if (dev == UINT_MAX) {
+        return 1;
+    }
+
+    pwm_mode_t pwm_mode;
+    switch(atoi(argv[2])) {
+        case(0):
+            pwm_mode = PWM_LEFT;
+            break;
+        case(1):
+            pwm_mode = PWM_RIGHT;
+            break;
+        case(2):
+            pwm_mode = PWM_CENTER;
+            break;
+        default:
+            printf("[pwm] Error: mode %d is not supported.\n", atoi(argv[2]));
+            return 1;
+    }
+
+    uint32_t pwm_freq = pwm_init(PWM_DEV(dev), pwm_mode,
+                                 atoi(argv[3]), atoi(argv[4]));
+    if (pwm_freq != 0) {
+        printf("[pwm] The pwm frequency is set to %ld\n", pwm_freq);
+        initiated |= (1 << dev);
+        return 0;
+    }
+
+    return 1;
+}
+
+int set(int argc, char**argv) {
+    if (argc != 4) {
+        printf("usage: %s <dev> <ch> <val>\n", argv[0]);
+        printf("\tdev: device by number between 0 and %d\n", PWM_NUMOF - 1);
+        puts("\tch: channel of device\n");
+        puts("\tval: duty cycle\n");
+        return 1;
+    }
+
+    unsigned dev = _get_dev(argv[1]);
+    if (dev == UINT_MAX) {
+        return 1;
+    }
+
+    if ((initiated & (1 << dev)) == 0) {
+        puts("[pwm] Error: pwm is not initiated.\n");
+        puts("Execute init function first.\n");
+        return 1;
+    }
+
+    if (atoi(argv[2]) >= pwm_channels(PWM_DEV(dev))) {
```

Comment:
better cast the `atoi` result to `unint8_t` here, so no compiler can complain about signed/unsigned comparison

Reply:
maybe event better: use a variable for the channel (`uint8_t chan = (uint8_t)atoi(argv[2]);`), as this variable is used more than once in the code below...